### PR TITLE
client: fix dirty caps might never be flushed

### PR DESCRIPTION
**Pull request by taodd**
issue tracker: http://tracker.ceph.com/issues/22546

1. in order to fix this issue, i have introduced an extra inode item - dirty_cap_item.
because, i don't want to share item with list "delayed_caps", it would make things complicated.
i can't find an elegant and correct way to share the items.
2. also replace the "cap_list" with "dirty_list" in class "Client",  flush_caps_sync would be more efficient, since it only need to check the dirty inodes instead of the whole inodes with caps.

Signed-off-by: dongdong tao <tdd21151186@gmail.com>

### CODE REVIEW

**Comment by ukernel** [comment]
how about this
<pre>
diff --git a/src/client/Client.cc b/src/client/Client.cc
index 1c2d66002e..d5377bd5e4 100644
--- a/src/client/Client.cc
+++ b/src/client/Client.cc
@@ -4206,6 +4206,9 @@ void Client::adjust_session_flushing_caps(Inode *in, MetaSession *old_s,  MetaSe
 void Client::flush_caps_sync()
 {
   ldout(cct, 10) << __func__ << dendl;
+
+  xlist<Inode*> tmp_list
+
   xlist<Inode*>::iterator p = delayed_caps.begin();
   while (!p.end()) {
     unsigned flags = CHECK_CAPS_NODELAY;
@@ -4213,6 +4216,8 @@ void Client::flush_caps_sync()
 
     ++p;
     delayed_caps.pop_front();
+    tmp_list.push_back(&in->cap_item);
+
     if (p.end() && cap_list.empty())
       flags |= CHECK_CAPS_SYNCHRONOUS;
     check_caps(in, flags);
@@ -4229,6 +4234,14 @@ void Client::flush_caps_sync()
       flags |= CHECK_CAPS_SYNCHRONOUS;
     check_caps(in, flags);
   }
+
+  p = tmp_list.begin();
+  while (!p.end()) {
+    Inode *in = *p;
+    ++p;
+    tmp_list.pop_front();
+    cap_list.push_back(&in->cap_item);
+  }
 }
 
 void Client::flush_caps(Inode *in, MetaSession *session, bool sync)
</pre>

**Comment by taodd** [comment]
thanks zheng, 
I have thought about to use a tmp list there too, that would be simply to fix this bug.
but two questions
1 if we use tmp list here, it's equivalent to another more clean way -> just push all the item from delay_caps to cap_list, then do the check in cap_list loop.
2 more important, do we really need to check all the inodes with caps here, i think only dirty inodes would be more efficient ?

**Comment by taodd** [comment]
please treat this more like a clean-up patch :)

**Comment by ukernel** [comment]
make sense

File: src/client/Client.cc [hunk 1]
```diff
@@ -4145,11 +4142,19 @@ void Client::force_session_readonly(MetaSession *s)
...
-  ldout(cct, 10) << "mark_caps_dirty " << *in << " " << ccap_string(in->dirty_caps) << " -> "
+  ldout(cct, 10) << __func__ << " " << *in << " " << ccap_string(in->dirty_caps) << " -> "
 	   << ccap_string(in->dirty_caps | caps) << dendl;
   if (caps && !in->caps_dirty())
     in->get();
   in->dirty_caps |= caps;
+  dirty_list.push_back(&in->dirty_cap_item);
+}
+
+void Client::mark_caps_clean(Inode *in)
```
Comment (batrick) [inline src/client/Client.cc]: Make this an `Inode` method.

**Comment by taodd** [comment]
@batrick  updated, i have also make "mark_caps_dirty" as a method of Inode,
i think it's more clean.
after this PR merged, i will try to clean the mark_caps_flushing, and it should be method of Inode too.
and another idea is that we might need to implement mark_caps_flushing_clean either.
i think it might clean much more code path.   
what do you think ?

File: src/client/Client.h [hunk 2]
```diff
@@ -730,6 +729,7 @@ class Client : public Dispatcher, public md_config_obs_t {
   void flush_mdlog_sync();
   void flush_mdlog(MetaSession *session);
   
+  xlist<Inode*> get_dirty_list() { return dirty_list; }
```
Comment (batrick) [inline src/client/Client.h]: should be `xlist<Inode *> &get_dirt_list()...`

**Comment by batrick** [comment]
> after this PR merged, i will try to clean the mark_caps_flushing, and it should be method of Inode too.
and another idea is that we might need to implement mark_caps_flushing_clean either.
i think it might clean much more code path.

I think leave it as is for now. There are too many dependencies now. Maybe as part of a more invasive reorganization we will revisit.

File: src/client/Client.h (hunk 2 above)
Comment (taodd) [reply src/client/Client.h]: thanks you
updated..

**Comment by taodd** [comment]
agree

2018年1月5日星期五，Patrick Donnelly <notifications@github.com> 写道：

**Pull request merged**